include/SFML/Audio/InputSoundFile.hpp

```diff
@@ -211,13 +212,13 @@ class SFML_AUDIO_API InputSoundFile : NonCopyable
     ////////////////////////////////////////////////////////////
     // Member data
     ////////////////////////////////////////////////////////////
-    SoundFileReader* m_reader;       //!< Reader that handles I/O on the file's format
-    InputStream*     m_stream;       //!< Input stream used to access the file's data
-    bool             m_streamOwned;  //!< Is the stream internal or external?
-    Uint64           m_sampleOffset; //!< Sample Read Position
-    Uint64           m_sampleCount;  //!< Total number of samples in the file
-    unsigned int     m_channelCount; //!< Number of channels of the sound
-    unsigned int     m_sampleRate;   //!< Number of samples per second
+    std::unique_ptr<SoundFileReader> m_reader;       //!< Reader that handles I/O on the file's format
+    InputStream*                     m_stream;       //!< Input stream used to access the file's data
+    bool                             m_streamOwned;  //!< Is the stream internal or external?
```

Comment:
`m_stream` could a unique_ptr with a custom deleter. See https://stackoverflow.com/a/24060783

Reply:
I've implemented that, thanks for the suggestion. Please take a look and let me know what you think :)

Reply:
Nice! The asymmetry between allocation with `new` and freeing with `std::default_delete` got me curious and I did some more tinkering. I got rid of the pointer casts in the process. See https://github.com/kimci86/SFML/commit/51574b4fce039625c9d87bb8b6b6851cbd30dfce.
Also I reordered some parts so that pointers are immediately deleted upon failure instead of waiting for the next call to `close()`. Feel to cherry-pick or take inspiration from what I did.

Reply:
@kimci86: that's a great improvement -- it is so much cleaner. Thanks! I want to cherry-pick, but I am not totally certain that removing those `close()` calls is correct -- `close()` also resets the sound file attributes to zero in case of failure, and it doesn't seem like we do that anymore if we fail to e.g. read a file. Can you confirm this will not cause any issue?

Reply:
`close()` is called at the beginning of each `openFromXXX` method and sound file attributes are not modified unless opening succeeds so it looks to me that it is not needed to reset them on failure.

Reply:
I am convinced, thanks! Will cherry pick ;)